quill-async/src/main/scala/io/getquill/sources/async/AsyncSource.scala

```diff
@@ -60,9 +54,9 @@ abstract class AsyncSource[D <: SqlIdiom, N <: NamingStrategy, C <: Connection](
       Await.result(pool.sendQuery(sql), Duration.Inf)
     }
 
-  def transaction[T](f: TransactionalExecutionContext => Future[T])(implicit ec: ExecutionContext) =
+  def transaction[T](f: AsyncSource[D, N, C] => Future[T]) =
     pool.inTransaction { c =>
-      f(TransactionalExecutionContext(ec, c))
+      f(this)
```

Comment:
This won't make the future run within the transaction. I suggest changing `AsyncSource` to have an abstract method `connection` and two sub-classes: one that receives `config` and returns a connection from the pool and another that will be created by `transaction` based on the single transactional connection provided by `inTransaction`.